packages/world/src/utils/InventoryUtils.sol

```diff
@@ -317,145 +380,188 @@ library InventoryUtils {
+
+        uint16 slot = uint16(i * SLOTS_PER_WORD + bitIndex);
+        if (InventorySlot._getObjectType(owner, slot) == objectType) {
+          tempSlots[count++] = slot;
+        }
+      }
+    }
 
-  function _removeFromOccupiedSlots(EntityId owner, uint16 slot) private {
-    uint16 occupiedIndex = InventorySlot._getOccupiedIndex(owner, slot);
-    uint256 last = Inventory._lengthOccupiedSlots(owner) - 1;
-    if (occupiedIndex < last) {
-      uint16 moved = Inventory._getItemOccupiedSlots(owner, last);
-      Inventory._updateOccupiedSlots(owner, occupiedIndex, moved);
-      InventorySlot._setOccupiedIndex(owner, moved, occupiedIndex);
+    // Resize array to actual count
+    uint16[] memory slots = new uint16[](count);
+    for (uint256 i = 0; i < count; i++) {
+      slots[i] = tempSlots[i];
     }
-    Inventory._popOccupiedSlots(owner);
+    return slots;
   }
 
-  // Add a slot to type slots - O(1)
-  function _addToTypeSlots(EntityId owner, ObjectType objectType, uint16 slot) private {
-    InventorySlot._setTypeIndex(owner, slot, uint16(InventoryTypeSlots._length(owner, objectType)));
-    InventoryTypeSlots._push(owner, objectType, slot);
-  }
+  function countObjectsOfType(EntityId owner, ObjectType objectType) internal view returns (uint256 total) {
+    uint256 bitmapLength = InventoryBitmap._length(owner);
+
+    for (uint256 i = 0; i < bitmapLength; i++) {
+      uint256 word = InventoryBitmap._getItem(owner, i);
+      while (word != 0) {
+        uint256 bitIndex = LibBit.ffs(word);
+        word &= word - 1;
 
-  // Remove a slot from type slots - O(1)
-  function _removeFromTypeSlots(EntityId owner, ObjectType objectType, uint16 slot) private {
-    uint16 typeIndex = InventorySlot._getTypeIndex(owner, slot);
-    uint256 last = InventoryTypeSlots._length(owner, objectType) - 1;
+        uint16 slot = uint16(i * SLOTS_PER_WORD + bitIndex);
+        InventorySlotData memory data = InventorySlot._get(owner, slot);
 
-    // If not the last element, swap with the last element
-    if (typeIndex < last) {
-      uint16 lastSlot = InventoryTypeSlots._getItem(owner, objectType, last);
-      InventoryTypeSlots._update(owner, objectType, typeIndex, lastSlot);
-      InventorySlot._setTypeIndex(owner, lastSlot, typeIndex);
+        if (data.objectType == objectType) {
+          total += data.amount;
+        }
+      }
     }
-    InventoryTypeSlots._pop(owner, objectType);
   }
 
-  function _replaceSlot(
-    EntityId owner,
-    uint16 slot,
-    ObjectType objectType,
-    EntityId entityId,
-    ObjectType newObjectType,
-    uint16 amount
-  ) internal {
-    _removeFromTypeSlots(owner, objectType, slot);
-    _addToTypeSlots(owner, newObjectType, slot);
-    InventorySlot._setEntityId(owner, slot, entityId);
-    InventorySlot._setObjectType(owner, slot, newObjectType);
-    InventorySlot._setAmount(owner, slot, amount);
+  function hasObjectType(EntityId owner, ObjectType objectType) internal view returns (bool) {
```

Comment:
Are these utils just here for reference? I dont think we're using them yet:
- hasObjectType
- countObjectsOfType
- getSlotsWithType
- getOccupiedSlotCount
- findObjectType

Reply:
ahh np, I dont mind them here, up to you